homer0/projext-plugin-webpack
# feat(project): update definitions when the bundle gets reloaded

### What does this PR do?

This is a follow up for #.

Variables from `.env` files and browser target configurations were being loaded and sent to the `webpack.DefinePlugin`, but the problem was that if a variable would change, even after triggering a bundle reload, its value wouldn't be updated on the target code nor its configuration.

Short version: Definitions are now refreshed every time the bundle gets reloaded.

And now, for the long a breaking version:

The first thing I did was to convert the `params.definitions` the engine moves around from an object into a function that would return the object.

Then, `webpack.DefinePlugin` allows for computed definitions as functions... but each key needs to be function, you can't have a single one that would return everything; so I created a webpack plugin that wraps `webpack.DefinePlugin` and supports a single function.

Finally, I replaced `webpack.DefinePlugin` on the browsers' configurations with the new plugin.

Now, **this is very important**: because using computed definitions without specifying a file that depends on the definition marks ALL files as "non-cachable", which makes the build slower, the build engine **only reloads the definitions when the target entry file changes**. Yes, it's not the best solution, but the alternative would be slower builds.

### How should it be tested manually?

In order to test this, you need to install `homer0/projext#next` and make sure you delete the copy of `projext` `npm`/`yarn` will install inside this package's `node_modules`.

Create a browser target, and `.env` file with a declared variable and configuration for the target that uses the variable.

Then run your target and check if the configuration has the right value. Update the variable on the `.env` file and "save the target entry file" (cmd/ctrl+s would trigger the rebuild), the configuration should have the new value.

And, of course...

```bash
yarn test
# or
npm test
```
